src/location-picker/location-picker.component.tsx

```diff
@@ -0,0 +1,182 @@
+import React from "react";
+import { always } from "kremling";
+import { debounce, isEmpty } from "lodash";
+import { createErrorHandler } from "@openmrs/esm-error-handling";
+import { Trans } from "react-i18next";
+import { LocationEntry } from "../types";
+import styles from "../styles.css";
+
+const CardHeader: React.FC = (props) => (
+  <div className={styles["card-header"]}>
+    <h2 className={`omrs-margin-8 omrs-margin-left-12`}>{props.children}</h2>
+  </div>
+);
+
+interface RadioInputProps {
+  option: LocationEntry;
+  current: string;
+  changeLocationData: (data: Partial<LocationDataState>) => void;
+}
+
+const RadioInput: React.FC<RadioInputProps> = ({
+  option,
+  current,
+  changeLocationData,
+}) => (
+  <div className="omrs-radio-button">
+    <input
+      id={option.resource.id}
+      type="radio"
+      name="location"
+      value={option.resource.id}
+      checked={current === option.resource.id}
+      onChange={(evt) =>
+        changeLocationData({ activeLocation: evt.target.value })
+      }
+    />
+    <label htmlFor={option.resource.id} className="omrs-padding-4">
+      {option.resource.name}
+    </label>
+  </div>
+);
+
+interface LocationDataState {
+  activeLocation: string;
+  locationResult: Array<LocationEntry>;
+}
+
+interface LocationPickerProps {
+  currentUser: string;
+  loginLocations: Array<LocationEntry>;
+  onChangeLocation(locationUuid: string): void;
+  searchLocations(query: string): Promise<Array<LocationEntry>>;
+}
+
+export default function LocationPicker(props: LocationPickerProps) {
+  const [locationData, setLocationData] = React.useState<LocationDataState>({
+    activeLocation: "",
+    locationResult: props.loginLocations,
+  });
+  const [searchTerm, setSearchTerm] = React.useState("");
+  const [isSubmitting, setIsSubmitting] = React.useState(false);
+
+  const searchTimeout = 300;
+
+  React.useEffect(() => {
+    if (isSubmitting) {
+      props.onChangeLocation(locationData.activeLocation);
+    }
+  }, [isSubmitting, locationData]);
+
+  React.useEffect(() => {
+    const ac = new AbortController();
+
+    if (props.loginLocations.length > 100) {
+      if (searchTerm) {
+        props.searchLocations(searchTerm).then((locationResult) => {
+          changeLocationData({
+            locationResult,
+          });
+        }, createErrorHandler());
+      }
+    } else if (searchTerm) {
+      filterList(searchTerm);
+    } else if (props.loginLocations !== locationData.locationResult) {
+      changeLocationData({ locationResult: props.loginLocations });
+    }
+
+    return () => ac.abort();
+  }, [searchTerm, props.loginLocations]);
```

Comment:
`filterList` is a dep

Reply:
No. Then this effect will *always* run as we set `filterList` in every render.

When we run this effect we only want to use the last (captured) version of it. We are not interested in future changes of `filterList`.

Reply:
Right, so shouldn't you wrap `filterList` in `useCallback`? https://reactjs.org/docs/hooks-faq.html#is-it-safe-to-omit-functions-from-the-list-of-dependencies

Reply:
`useCallback` as the name suggests should be (mainly) used for callbacks placed in other components. That way these components do not re-render necessarily. Otherwise, its the same as with the effect. They would always re-render since the function ref changed (as it was re-created in the re-render of the parent component).

Again, we are here re-creating the function `filterList` on every render. We don't place a dep on this as we just take the most recent one when the effect is run. This is all perfectly fine with the current algorithm. Unfortunately, the linting tools are insufficient for this case. They only cover the 80-90% case.

Reply:
Oh yeah, you're right! I'm starting to understand the `extensive-deps` lint rule better. Based on [that one issue comment](https://github.com/facebook/react/issues/14920#issuecomment-471070149) from Dan Abramov, "Functions in the same component but defined outside of the effect", the lint rule should accept this usage, also (I haven't checked the linter output). Thanks!